optuna/samplers/_tpe/parzen_estimator.py

```diff
@@ -54,424 +45,229 @@ def __init__(
         predetermined_weights: Optional[np.ndarray] = None,
     ) -> None:
+        if parameters.consider_prior:
+            if parameters.prior_weight is None:
+                raise ValueError("Prior weight must be specified when consider_prior==True.")
+            elif parameters.prior_weight <= 0:
+                raise ValueError("Prior weight must be positive.")
 
         self._search_space = search_space
-        self._parameters = parameters
-        self._n_observations = next(iter(observations.values())).size
-        if predetermined_weights is not None:
-            assert self._n_observations == len(predetermined_weights)
-        self._weights = self._calculate_weights(predetermined_weights)
-
-        self._low: Dict[str, Optional[float]] = {}
-        self._high: Dict[str, Optional[float]] = {}
-        self._q: Dict[str, Optional[float]] = {}
-        for param_name, dist in search_space.items():
-            if isinstance(dist, distributions.CategoricalDistribution):
-                low = high = q = None
+
+        transformed_observations = self._transform(observations)
+
+        assert predetermined_weights is None or len(transformed_observations) == len(
+            predetermined_weights
+        )
+        weights = (
+            predetermined_weights
+            if predetermined_weights is not None
+            else self._call_weights_func(parameters.weights, len(transformed_observations))
+        )
+        if parameters.consider_prior or len(transformed_observations) == 0:
+            if len(transformed_observations) == 0:
+                prior_weight = 1.0
             else:
-                low, high, q = self._calculate_parzen_bounds(dist)
-            self._low[param_name] = low
-            self._high[param_name] = high
-            self._q[param_name] = q
-
-        # `_low`, `_high`, `_q` are needed for transformation.
-        observations = self._transform_to_uniform(observations)
-
-        # Transformed `observations` might be needed for following operations.
-        self._sigmas0 = self._precompute_sigmas0(observations)
-
-        self._mus: Dict[str, Optional[np.ndarray]] = {}
-        self._sigmas: Dict[str, Optional[np.ndarray]] = {}
-        self._categorical_weights: Dict[str, Optional[np.ndarray]] = {}
-        categorical_weights: Optional[np.ndarray]
-        for param_name, dist in search_space.items():
-            param_observations = observations[param_name]
-            if isinstance(dist, distributions.CategoricalDistribution):
-                mus = sigmas = None
-                categorical_weights = self._calculate_categorical_params(
-                    param_observations, param_name
+                assert parameters.prior_weight is not None
+                prior_weight = parameters.prior_weight
+            weights = np.append(
+                weights,
+                [1.0 if len(transformed_observations) == 0 else prior_weight],
+            )
```

Comment:
If `parameters.prior_weight` must be provided when `parameters.consider_prior = True`, the following simplification is possible.
```python
prior_weight = 1.0 if len(transformed_observations) == 0 else parameters.prior_weight
if prior_weight is not None:
    weights = np.apend(weights, prior_weight)
```

Reply:
When `parameters.prior_weight is not None` and `parameters.consider_prior == False`, this code wrongly adds prior.

Reply:
Personally, I think we could just set `prior_weight == 0` when `consider_prior == False`. However, that will be another medium-sized refactoring, so let me tackle that in a follow-up.

Reply:
So, how about the following?

```python
prior_weight = 1.0 if len(transformed_observations) == 0 else parameters.prior_weight
if prior_weight is not None and parameters.consider_prior:
    weights = np.apend(weights, prior_weight)
```

Reply:
This is also wrong because when `len(transformed_observations) == 0` and `parameters.consider_prior == False`, we *should* add prior.

Reply:
I could have just written 
```
weights = np.append(
                weights,
                [1.0 if len(transformed_observations) == 0 else parameters.prior_weight],
            )
```
but the type checker didn't let me.

Reply:
I simplified the logic here. PTAL.